### ISSUE TITLE
Ecosystem Upgrade: Composer Plugin Readiness for 2.0

### ISSUE THREAD

**Issue by Seldaek** [composer/composer#8726]
### Update September: See the [list of plugins still lacking support](https://github.com/composer/composer/issues/8726#issuecomment-691097291) below if you want to help

As we are getting Composer 2.0 closer and closer to a stable release,
and would like to get more people involved with testing it, it is
quite important to get the most common plugins to support our 2.0
branch. Otherwise many users will not be able to try the latest
Composer snapshots when we switch them to 2.0.

If you maintain a Composer plugin and would like to help, here is how:

- Get a composer 2.0 build by running `composer self-update --snapshot`

- Set up a trial project requiring your plugin from a local path repo
so you can easily work on it and iterate, e.g.

```
    "repositories": [
        {"type":"path", "url": "../../my-plugin"}
    ],
```

- Change your plugin's composer.json to require `"composer-plugin-api":
"^1.0 || ^2.0"` (to begin with, adjust to ^2.0 only if you can't make this work)

- Update your main plugin entry point class to implement the new
PluginInterface methods, see e.g.
https://github.com/composer/installers/commit/d2a98321ab66945396ef147387f93eee4c05089f#diff-e236c450b11f27e4bb81a5069bc619bc

- Check out the [UPGRADE guide](https://github.com/composer/composer/blob/master/UPGRADE-2.0.md#for-integrators-and-plugin-authors) for any other changes which might be relevant to you.

- Run `php composer.phar update` in that trial project to play with
your plugin using Composer 2, and fix any issues you run into. If you require other plugins which are not compatible yet, you can bypass that using `--ignore-platform-reqs` as it will ignore the invalid composer-plugin-api requirements.

- If at all possible, releasing a version which supports both
composer-plugin-api ^1.0 and ^2.0 is the best option really as that
will allow people to use whichever composer version within a project,
and makes the transition smoother. I am sure the more involved plugins
will not be able to do this but I think for most it should be
possible.

- Finally, report issues to us at
https://github.com/composer/composer/issues if anything is broken or
unclear, we'll try to help out ASAP.

Thanks for your attention and help getting the Composer v2 ecosystem up to speed!

**Comment by jrfnl** [comment]
@Seldaek Suggestion: make this a "pinned" issue ;-)

**Comment by Seldaek** [comment]
Published an [UPGRADE guide](https://github.com/composer/composer/blob/master/UPGRADE-2.0.md#for-integrators-and-plugin-authors) now which might help. I think it covers most relevant changes, but we did change more than that for sure in internal classes.

**Comment by cweagans** [comment]
Do you have a target date for the Composer 2 release? Or even a target month or quarter?

It'd be great to know when this work _needs_ to be completed.

**Comment by Seldaek** [comment]
@cweagans we don't have a strict timeline, but looking at the 2.0-core milestone, I'd say we are getting quite close. So I'm sure we'll do a alpha release soonish, then barring any onslaught of issues I'm hoping for a stable sometime late May or latest June.. If you have issues porting plugins please reach out, we're happy to help.

**Comment by cweagans** [comment]
Sounds great! Thanks for the context. That helps a lot!

**Comment by samdark** [comment]
Have an issue with Composer 2 that isn't reproduce-able with Composer 1: https://github.com/composer/composer/issues/8849

**Comment by greg-1-anderson** [comment]
FYIO, the [Drupal Composer plugins are now compatible with Composer 2](https://www.drupal.org/project/drupal/issues/3126566). Both Composer 1 and Composer 2 plugin APIs are supported.

These will be available in the stable releases 8.9.0 and 9.0.0; at the moment, they are only available in dev releases.

**Comment by R-cloud-pixel** [comment]
composer self - update --snapshot      composer -plugin    v1 v2    - - ignore - platform - req.

**Comment by BonBonSlick** [comment]
@R-cloud-pixel please do no suggest a temp or bad solutions  
running  `- - ignore - platform - req.` might lead to bigger issues

**Comment by R-cloud-pixel** [comment]
feature/composer-2      New feature or request    Multiple projects use different versions of packages and composer- plugin-2.0.0 as a result composer/composer

**Comment by R-cloud-pixel** [comment]
Ecosystem Upgrade: Composer plugin Readiness for 2.0 Issue #8726 composer/composer                         #8895 Composer v2 is faster for some projects but not for the others.                                                       #8910Downgrade transitive dependency to satisfy deps on "require".                                                        #8891 Print warning for old PHP versions.

**Comment by R-cloud-pixel** [comment]
#8653 Composer 1.9.3 first installs, then downgrades.

**Comment by R-cloud-pixel** [comment]
#8910 Downgrade transitive dependency to satisfy deps on "require".

**Comment by emodric** [comment]
Hi,

I have an issue when using https://github.com/emodric/ezpublish-legacy-installer/tree/composer2 to install eZ Publish Legacy extensions. The plugin does nothing more than change the path of the installed packages. They are installed into `ezpublish_legacy/extension` folder, and this works fine, however, on `composer update`, I get the following issue: `Expected empty path to extract ezsystems/ezwt-ls-extension-5.3.10.0 into but directory exists: ezpublish_legacy/extension/ezwt`.

What am I doing wrong?

@Seldaek maybe?

**Comment by Seldaek** [comment]
One new change which is important for custom installers extending from LibraryInstaller:

You should be aware that in Composer 2 since https://github.com/composer/composer/pull/8952 it MAY return PromiseInterface instances when calling parent::install/update/uninstall/installCode/removeCode. See [composer/installers](https://github.com/composer/installers/commit/5006d0c28730ade233a8f42ec31ac68fb1c5c9bb) for an example of how to handle this best.

**Comment by Seldaek** [comment]
@emodric so for example in that plugin of yours that'd be https://github.com/emodric/ezpublish-legacy-installer/blob/composer2/src/eZ/Publish/Composer/LegacyKernelInstaller.php#L71  and https://github.com/emodric/ezpublish-legacy-installer/blob/composer2/src/eZ/Publish/Composer/LegacyKernelInstaller.php#L114 where you need to make sure you handle the promises, otherwise you might end up with a race condition. 

I'm not sure if that explains the issue you're seeing tho.. if you still see it after fixing that please open a new issue with more details on how I can reproduce.

**Comment by emodric** [comment]
@Seldaek Sorry I was not specific, the installer in use is actually https://github.com/emodric/ezpublish-legacy-installer/blob/composer2/src/eZ/Publish/Composer/LegacyExtensionInstaller.php, which does not override `install/uninstall` methods and only changes the install path. I will open an issue as I've narrowed down the cause and the workflow that leads to the issue.

**Comment by Seldaek** [comment]
Ok but what I said above still applies, code overriding LibraryInstaller methods other than getInstallPath will most likely cause problems at some point if not updated correctly for v2 support.

**Comment by emodric** [comment]
@Seldaek Thanks for the tips! The installer you pointed to does cause issues, but a separate ones and unrelated to the one I posted.

**Comment by staabm** [comment]
@Seldaek can you foresee problems when different people work on a single project and some of them using composer 2.x while others still use 1.x?

**Comment by Seldaek** [comment]
Probably minor changes in lock file every time someone runs an update/require if you play ping pong between versions. It shouldn't cause issues per se, but can get annoying in the diffs.

Aside from that I don't think there are major issues as long as your plugins are compatible with both versions.

Note that you can easily add a require for `composer-runtime-api >=2`  if you want to force everyone to use v2.

**Comment by jrfnl** [comment]
@Seldaek Is plugin readiness being tracked somewhere ?
I.e. is there a list somewhere where Composer plugins can indicate that the plugin is ready for Composer 2.0, including as-of which version of the plugin Composer 2.0 is supported ?

**Comment by Seldaek** [comment]
A little update here on the state of things..

Of the 1345 plugins listed on packagist.org, only 110 (8%) have support for composer 2 listed on their default branch. Not so great but there's lots of old stuff in there.

Of the top 100 (by downloads) plugins, 39% have support, which is better but there's still room for improvement there.

The following are the 61 plugins from the top 100 which do not yet support v2:

- [drupal-composer/drupal-scaffold](https://packagist.org/packages/drupal-composer/drupal-scaffold) 7066497 downloads
- [drupal/console-extend-plugin](https://packagist.org/packages/drupal/console-extend-plugin) 6318798 downloads
- [wikimedia/composer-merge-plugin](https://packagist.org/packages/wikimedia/composer-merge-plugin) 6238665 downloads
- [magento/magento-composer-installer](https://packagist.org/packages/magento/magento-composer-installer) 6115240 downloads
- [robloach/component-installer](https://packagist.org/packages/robloach/component-installer) 5386631 downloads
- [fxp/composer-asset-plugin](https://packagist.org/packages/fxp/composer-asset-plugin) 3770419 downloads
- [zaporylie/composer-drupal-optimizations](https://packagist.org/packages/zaporylie/composer-drupal-optimizations) 2586180 downloads
- [zendframework/zend-component-installer](https://packagist.org/packages/zendframework/zend-component-installer) 1584258 downloads
- [codeception/c3](https://packagist.org/packages/codeception/c3) 1573332 downloads
- [grasmash/drupal-security-warning](https://packagist.org/packages/grasmash/drupal-security-warning) 1180741 downloads
- [drupal/core-composer-scaffold](https://packagist.org/packages/drupal/core-composer-scaffold) 1051989 downloads
- [raulfraile/ladybug-installer](https://packagist.org/packages/raulfraile/ladybug-installer) 921308 downloads
- [g1a/composer-test-scenarios](https://packagist.org/packages/g1a/composer-test-scenarios) 845874 downloads
- [rvtraveller/qs-composer-installer](https://packagist.org/packages/rvtraveller/qs-composer-installer) 766233 downloads
- [muglug/package-versions-56](https://packagist.org/packages/muglug/package-versions-56) 755593 downloads
- [topthink/think-installer](https://packagist.org/packages/topthink/think-installer) 707423 downloads
- [wp-cli/autoload-splitter](https://packagist.org/packages/wp-cli/autoload-splitter) 651569 downloads
- [laminas/laminas-dependency-plugin](https://packagist.org/packages/laminas/laminas-dependency-plugin) 639980 downloads
- [neos/composer-plugin](https://packagist.org/packages/neos/composer-plugin) 592093 downloads
- [socialengine/composer-symlinker](https://packagist.org/packages/socialengine/composer-symlinker) 589738 downloads
- [dg/composer-cleaner](https://packagist.org/packages/dg/composer-cleaner) 565982 downloads
- [silverstripe/vendor-plugin](https://packagist.org/packages/silverstripe/vendor-plugin) 531756 downloads
- [philippbaschke/acf-pro-installer](https://packagist.org/packages/philippbaschke/acf-pro-installer) 518340 downloads
- [simplyadmire/composer-plugins](https://packagist.org/packages/simplyadmire/composer-plugins) 487532 downloads
- [silverstripe/recipe-plugin](https://packagist.org/packages/silverstripe/recipe-plugin) 486314 downloads
- [webimpress/composer-extra-dependency](https://packagist.org/packages/webimpress/composer-extra-dependency) 452121 downloads
- [drupal-composer/preserve-paths](https://packagist.org/packages/drupal-composer/preserve-paths) 450280 downloads
- [davidbarratt/custom-installer](https://packagist.org/packages/davidbarratt/custom-installer) 417970 downloads
- [neronmoon/scriptsdev](https://packagist.org/packages/neronmoon/scriptsdev) 388361 downloads
- [magento-hackathon/magento-composer-installer](https://packagist.org/packages/magento-hackathon/magento-composer-installer) 356882 downloads
- [aydin-hassan/magento-core-composer-installer](https://packagist.org/packages/aydin-hassan/magento-core-composer-installer) 355243 downloads
- [mouf/nodejs-installer](https://packagist.org/packages/mouf/nodejs-installer) 351968 downloads
- [vaimo/composer-patches](https://packagist.org/packages/vaimo/composer-patches) 337423 downloads
- [mouf/mouf-installer](https://packagist.org/packages/mouf/mouf-installer) 301909 downloads
- [zfcampus/zf-asset-manager](https://packagist.org/packages/zfcampus/zf-asset-manager) 294821 downloads
- [mouf/html.utils.weblibrarymanager.component-installer](https://packagist.org/packages/mouf/html.utils.weblibrarymanager.component-installer) 275577 downloads
- [contao-community-alliance/composer-plugin](https://packagist.org/packages/contao-community-alliance/composer-plugin) 263978 downloads
- [helhum/class-alias-loader](https://packagist.org/packages/helhum/class-alias-loader) 250859 downloads
- [drupal-composer/drupal-paranoia](https://packagist.org/packages/drupal-composer/drupal-paranoia) 243923 downloads
- [jkribeiro/drupal-composer-paranoia-acquia](https://packagist.org/packages/jkribeiro/drupal-composer-paranoia-acquia) 235654 downloads
- [thadafinser/package-info](https://packagist.org/packages/thadafinser/package-info) 229218 downloads
- [edisonlabs/merge-yaml](https://packagist.org/packages/edisonlabs/merge-yaml) 216105 downloads
- [zendframework/zend-skeleton-installer](https://packagist.org/packages/zendframework/zend-skeleton-installer) 210160 downloads
- [oxid-esales/oxideshop-unified-namespace-generator](https://packagist.org/packages/oxid-esales/oxideshop-unified-namespace-generator) 200177 downloads
- [oxid-esales/oxideshop-composer-plugin](https://packagist.org/packages/oxid-esales/oxideshop-composer-plugin) 188752 downloads
- [eloquent/composer-npm-bridge](https://packagist.org/packages/eloquent/composer-npm-bridge) 186100 downloads
- [ec-cube/plugin-installer](https://packagist.org/packages/ec-cube/plugin-installer) 173873 downloads
- [nodejs-php-fallback/nodejs-php-fallback](https://packagist.org/packages/nodejs-php-fallback/nodejs-php-fallback) 169999 downloads
- [indigophp/doctrine-annotation-autoload](https://packagist.org/packages/indigophp/doctrine-annotation-autoload) 165323 downloads
- [metasyntactical/composer-plugin-license-check](https://packagist.org/packages/metasyntactical/composer-plugin-license-check) 165187 downloads
- [derhasi/composer-preserve-paths](https://packagist.org/packages/derhasi/composer-preserve-paths) 159671 downloads
- [drupal/core-vendor-hardening](https://packagist.org/packages/drupal/core-vendor-hardening) 157592 downloads
- [vaimo/binary-chromedriver](https://packagist.org/packages/vaimo/binary-chromedriver) 157346 downloads
- [hostnet/path-composer-plugin-lib](https://packagist.org/packages/hostnet/path-composer-plugin-lib) 146427 downloads
- [ffraenz/private-composer-installer](https://packagist.org/packages/ffraenz/private-composer-installer) 145257 downloads
- [hhvm/hhvm-autoload](https://packagist.org/packages/hhvm/hhvm-autoload) 143315 downloads
- [naderman/composer-aws](https://packagist.org/packages/naderman/composer-aws) 137304 downloads
- [franzl/studio](https://packagist.org/packages/franzl/studio) 128176 downloads
- [thecodingmachine/discovery](https://packagist.org/packages/thecodingmachine/discovery) 126843 downloads
- [netresearch/composer-patches-plugin](https://packagist.org/packages/netresearch/composer-patches-plugin) 120379 downloads

I am sure some already have some update in progress, and some might not be relevant anymore.. But if you want to help, feel free to go through the list and make sure there is at least an issue open to track this, or send a PR if you are familiar enough with the plugin to do that safely.

**Comment by jrfnl** [comment]
Of the above listed packages, the following ones are marked as abandoned, so no update should be expected:

* [robloach/component-installer](https://packagist.org/packages/robloach/component-installer)
    The author suggests using the [oomphinc/composer-installers-extender](https://packagist.org/packages/oomphinc/composer-installers-extender) package instead.
    This alternative is marked as compatible with Composer 2.0.
* [zendframework/zend-component-installer](https://packagist.org/packages/zendframework/zend-component-installer)
    The author suggests using the [laminas/laminas-component-installer](https://packagist.org/packages/laminas/laminas-component-installer) package instead. 
    This alternative is marked as compatible with Composer 2.0.
* [wp-cli/autoload-splitter](https://packagist.org/packages/wp-cli/autoload-splitter)
    Abandoned without alternative suggested, however the authors indicated that the package is no longer needed as of WPCLI 2.0.
* [simplyadmire/composer-plugins](https://packagist.org/packages/simplyadmire/composer-plugins)
    Abandoned without alternative suggested.
    From personal experience, I can recommend switching to the [dealerdirect/phpcodesniffer-composer-installer](https://packagist.org/packages/dealerdirect/phpcodesniffer-composer-installer) package for this functionality, which is compatible with Composer 2.0.
* [zfcampus/zf-asset-manager](https://packagist.org/packages/zfcampus/zf-asset-manager)
    The author suggests using the [laminas-api-tools/api-tools-asset-manager](https://packagist.org/packages/laminas-api-tools/api-tools-asset-manager) package instead.
    This alternative is marked as compatible with Composer 2.0.
* [helhum/class-alias-loader](https://packagist.org/packages/helhum/class-alias-loader)
    The author suggests using the [typo3/class-alias-loader](https://packagist.org/packages/typo3/class-alias-loader) package instead.
    This alternative is marked as compatible with Composer 2.0.
* [zendframework/zend-skeleton-installer](https://packagist.org/packages/zendframework/zend-skeleton-installer)
    The author suggests using the [laminas/laminas-skeleton-installer](https://packagist.org/packages/laminas/laminas-skeleton-installer) package instead. 
    This alternative is marked as compatible with Composer 2.0.
* [derhasi/composer-preserve-paths](https://packagist.org/packages/derhasi/composer-preserve-paths)
    The author suggests using the [drupal-composer/preserve-paths](https://packagist.org/packages/drupal-composer/preserve-paths) package instead.
    This alternative is NOT compatible (yet) with Composer 2.0.

The following packages might be "false positives", as it looks like they indicate that the package *is* compatible with Composer 2.0 (`composer-plugin-api: ^1 || ^2` in `require`). It might be that they only still need update the `composer/composer` constraints in `require-dev`:
* [drupal/core-composer-scaffold](https://packagist.org/packages/drupal/core-composer-scaffold)
* [neos/composer-plugin](https://packagist.org/packages/neos/composer-plugin)
* [drupal/core-vendor-hardening](https://packagist.org/packages/drupal/core-vendor-hardening)

In other words:
> The following are the ~~61~~ **50** plugins from the top 100 which do not yet support v2:

**Comment by Seldaek** [comment]
Thanks @jrfnl for going through! The 3 false positives are because my sql query only looked at the default branch, and they seem to have it set to an older release branch. 

Anyway I updated the list above to remove all those, so nobody else spends time looking at them.

**Comment by greg-1-anderson** [comment]
drupal/core-composer-scaffold and drupal/core-vendor-hardening only need Composer 1 for testing. We can't go up to Composer here until we also bump composer/semver in Drupal core. These plugins work with Composer 2, though, so this should not be an issue.

drupal-composer/drupal-scaffold is deprecated; everyone should switch to drupal/core-composer-scaffold. Note that it is possible to use drupal/core-composer-scaffold from Drupal ^8.9 to install really old versions of Drupal 8, back to Drupal 8.1.0 should work, although technically only 8.7.0+ is supported.

I've got [one more plugin](https://packagist.org/packages/g1a/composer-test-scenarios) from that list that I can update to Composer 2 shortly.

**Comment by mcaskill** [comment]
Hi,

I think `UPGRADE-2.0.md` should explain how locking/updating/installing has changed in Composer 2.
I've noticed across a few plugins that developers not fully aware of the rather drastic change in how Composer operates. Roughly speaking:

- Composer v1: Composer resolves dependencies, iterates packages (while dispatching `PRE_PACKAGE_*` before `PRE_FILE_DOWNLOAD`), then finally writes the lock file.
- Composer v2: Composer resolves dependencies, writes the lock file, downloads the packages (while dispatching `PRE_FILE_DOWNLOAD`), then iterates packages (while dispatching `PRE_PACKAGE_*`).

In particular, people are assuming that `PRE_OPERATIONS_EXEC` is a replacement for `PRE_DEPENDENCIES_SOLVING` since they are dispatched in a similar-looking routine. The closest event to the latter would actually be `PRE_POOL_CREATE`.

Thanks,

**Comment by Seldaek** [comment]
@mcaskill good point thanks, I added detailed notes and a debug mode for events in https://github.com/composer/composer/commit/3c25d180e63f9e0dd2f41c3e1a46a6027c0ba80e

**Comment by Seldaek** [comment]
Closing here as Composer 2.0 is out, not that the migration work is over but it probably never will be :)
